Fix a few issues with database upgrade

peppy opened the pull request.

peppy committed:
Hidden cannot be null

peppy committed:
Keep trying until delete succeeds

Turns out it can fail if file handles are still open.

peppy committed:
Allow recovery from a very broken database

smoogipoo commented on osu.Game/OsuGameBase.cs:
```diff
@@ -97,8 +97,18 @@ private void load()
+                        }
+                    }
+                    catch { }
+                }
```
This is suuuuuuuuuper dodgy.

peppy replied:
suggestions?

peppy committed:
Add maximum try count before bailing

peppy committed:
Add more logging output

peppy committed:
Ignore include-ignore warnings for now

smoogipoo commented on osu.Game/OsuGameBase.cs:
```diff
@@ -173,6 +163,42 @@ private void load()
+                    catch
+                    {
+                    }
+                }
```
I mean first of all the point of using `using ()` is to ensure that the objects and their underlying resources are deterministically closed. If EnsureDeleted is failing while all the resources from the previous context are supposedly closed, I want to know why.
Please provide the exception this is failing with.

Secondly if EnsureDeleted fails for these 20 attempts we're still left with an unusable database. What do?

Should we start using retries any time file access is involved?

peppy replied:
fails with file-in-use. unsure why, possibly sqlite holding an open file handle slightly longer than it needs? there's many layers interacting here which can lead to these kinds of issues (EF -> EF-Database -> sqlite native)

if it fails, it will attempt again the next time the user runs the app. keep in mind this isn't to do with just sqlite-net migration. this recoveres from *any* database bad states, including EF migrations, so "removing migration" is not an option.

peppy committed:
Output the inner exception to the log when possible

peppy committed:
Rely on storage.DeleteDatabase for guaranteed database deletion

Relies on https://github.com/ppy/osu-framework/pull/1100 being merged for most effectiveness.

peppy committed:
Merge remote-tracking branch 'upstream/master' into db-migration-fixes
